xarray/core/accessor_str.py

```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            return self._obj.copy().expand_dims({dim: 0}, -1)
+
+        f_count = lambda x: max(len(x.rsplit(sep, maxsplit)), 1)
+        maxsplit = self._apply(f_count, dtype=np.int_).max().data.tolist() - 1
+
+        def _dorsplit(mystr, sep=sep, maxsplit=maxsplit, dtype=self._obj.dtype):
+            res = mystr.rsplit(sep, maxsplit)
+            if len(res) < maxsplit + 1:
+                res = [""] * (maxsplit + 1 - len(res)) + res
+            return np.array(res, dtype=dtype)
+
+        # dtype MUST be object or strings can be truncated
+        # See: https://github.com/numpy/numpy/issues/8352
+        return self._apply(
+            _dorsplit,
+            dtype=np.object_,
+            output_core_dims=[[dim]],
+            output_sizes={dim: maxsplit},
+        ).astype(self._obj.dtype.kind)
+
+    def get_dummies(
+        self,
+        dim: Hashable,
+        sep: str = "|",
+    ) -> Any:
+        """
+        Return DataArray of dummy/indicator variables.
+
+        Each string in the DataArray is split at `sep`.
+        A new dimension is created with coordinates for each unique result,
+        and the corresponding element of that dimension is `True` if
+        that result is present and `False` if not.
+
+        Parameters
+        ----------
+        dim : Hashable
+            Name for the dimension to place the results in.
+        sep : str, default `"|"`.
+            String to split on.
+
+        Returns
+        -------
+        dummies : array of bool
+
+        Examples
+        --------
+        Create a string array
+
+        >>> values = xr.DataArray(
+        ...     [
+        ...         ["a|ab~abc|abc", "ab", "a||abc|abcd"],
+        ...         ["abcd|ab|a", "abc|ab~abc", "|a"],
+        ...     ],
+        ...     dims=["X", "Y"],
+        ... )
+
+        Extract dummy values
+
+        >>> values.str.get_dummies(dim="ZZ")
+        <xarray.DataArray (X: 2, Y: 3, ZZ: 5)>
+        array([[[ True, False,  True, False,  True],
+                [False,  True, False, False, False],
+                [ True, False,  True,  True, False]],
+        <BLANKLINE>
+               [[ True,  True, False,  True, False],
+                [False, False,  True, False,  True],
+                [ True, False, False, False, False]]])
+        Coordinates:
+          * ZZ       (ZZ) <U6 'a' 'ab' 'abc' 'abcd' 'ab~abc'
+        Dimensions without coordinates: X, Y
+
+        See Also
+        --------
+        pandas.Series.str.get_dummies
+        """
+        if not self._obj.size:
+            return self._obj.copy().expand_dims({dim: 0}, -1)
+
+        sep = self._obj.dtype.type(sep)
+        f_set = lambda x: set(x.split(sep)) - {self._obj.dtype.type("")}
+        setarr = self._apply(f_set, dtype=np.object_)
+        vals = sorted(reduce(set_union, setarr.data.ravel()))
+
+        f = lambda x: np.array([val in x for val in vals], dtype=np.bool_)
+        res = self._apply(
+            f,
+            obj=setarr,
```

Comment:
As far as I can see this is the only place you use `obj` - could this be refactored?

Reply:
Do you have a way to do that which wouldn't require considerably more code and complexity?